[clr-interp] Implement intrinsics for startup path

kotlarmilos opened the pull request:
## Description

The goal is to identify and implement additional missing intrinsics used during startup.

This PR implements missing intrinsics for the startup path. It builds on changes from https://github.com/dotnet/runtime/pull/116769 and https://github.com/dotnet/runtime/pull/116924 which fixed conv opcodes and introduced scaffolding for intrinsic handling. This PR is a follow-up to mentioned PRs and will not be merged until those are merged.

## Changes
 - Implemented StaticsHelpers.VolatileReadAsByref
 - Implemented Interlocked.CompareExchange
 - Implemented RuntimeHelpers.GetMethodTable
 - Implemented MemoryMarshal.GetArrayDataReference
 - Implemented RuntimeHelpers.IsReferenceOrContainsReferences

## Tasks
 - [x] Add tests

Run a sample desktop app using the following environment variables:
```
 "DOTNET_Interpreter" : "*!*",
 "DOTNET_ReadyToRun" : "0"
```

kotlarmilos committed:
Test startup path

kotlarmilos committed:
Merge branch 'main' into feature/coreclr-interp-startup-path

kotlarmilos committed:
Avoid memset in newobj.vt

kotlarmilos committed:
Merge branch 'feature/coreclr-interp-startup-path' of github.com:kotlarmilos/runtime into feature/coreclr-interp-startup-path

kotlarmilos committed:
Add support for FastPollGC intrinsic

kotlarmilos committed:
Merge branch 'main' into feature/coreclr-interp-startup-path

kotlarmilos committed:
Resolve merge conflicts

kotlarmilos committed:
Implement COMPARE_EXCHANGE macro

kotlarmilos commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1944,8 +1944,8 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
 
-                    // clear the valuetype
-                    memset(vtThis, 0, vtSize);
+                    // Interpreter-TODO: Clear the valuetype memory for GC encoding
```
@davidwrighton In what scenario is this needed?

janvorli replied:
This depends on how we resolve the GC reporting around calls where the return value overlaps with arguments. We cannot zero the memory unless we ensure that it doesn't overlap with the arguments. 
@kg already has some ideas on that.

BrzVlad replied:
This should never overlap with arguments. I don't understand why are we removing this zeroing as part of this PR. From what I understand from the spec, newobj needs to zero the new object/vt. I think that ctors fully initializing the object is just a C# requirement.

kotlarmilos replied:
This is encountered at the startup path in `newobj.vt`:
```
Compiled method: .System.RuntimeTypeHandle:GetInstantiationInternal()
Locals size 112
IR_0000: initlocals     [nil <- nil], 16,32
IR_0003: safepoint      [nil <- nil],
IR_0004: ldc.i8.0       [48 <- nil],
IR_0006: mov.8          [16 <- 48],
IR_0009: mov.8          [64 <- 0],
IR_000c: call           [48 <- 64], .System.RuntimeTypeHandle:GetNativeHandle()
IR_0010: mov.vt         [24 <- 48], 8
IR_0014: ldloca         [72 <- nil], 24
IR_0017: newobj.vt      [64 <- 64], .System.Runtime.CompilerServices.QCallTypeHandle:.ctor(byref)
```
The `QCallTypeHandle` is 16 bytes and the offset to the first argument is only 8 bytes. I still don't have a minimal repro, so let's fix this in a follow-up. The change has been reverted.

kotlarmilos committed:
Remove comment

kotlarmilos committed:
Merge branch 'main' into feature/coreclr-interp-startup-path

kotlarmilos committed:
Resolve merge conflicts

janvorli commented on src/coreclr/interpreter/eeinterp.cpp:
```diff
@@ -106,6 +106,11 @@ CorJitResult CILInterp::compileMethod(ICorJitInfo*         compHnd,
 
     try
     {
+        const char *className = NULL;
```
We don't need this, we already have `DOTNET_InterpList` that enables dumping all the compiled methods.

janvorli commented on src/coreclr/interpreter/intops.def:
```diff
@@ -403,6 +403,12 @@ OPDEF(INTOP_GC_COLLECT, "gc.collect", 1, 0, 0, InterpOpNoArgs)
 OPDEF(INTOP_LOAD_FRAMEVAR, "load.framevar", 2, 1, 0, InterpOpNoArgs)
 
+// Intrinsics
+OPDEF(INTOP_METHODTABLE, "methodtable", 3, 1, 1, InterpOpInt)
```
can we name this INTOP_GETMETHODTABLE / getmethodtable for better clarity?

BrzVlad replied:
Do we even need this opcode ? Shouldn't this be just a `LDIND_I` with offset 0 ?

kotlarmilos replied:
No need, replaced with LDIND_I. Thanks.

janvorli commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -2180,6 +2180,193 @@ int32_t InterpCompiler::GetDataItemIndexForHelperFtn(CorInfoHelpFunc ftn)
+
+            // Interpreter-FIXME: Implement conversion before INTOP_COMPARE_EXCHANGE to avoid multiple opcodes
+            int32_t opcode;
+            switch (retType)
```
Mono does it the same way - having multiple IR opcodes. We could possibly use a single opcode and pass in size, but then we would have to have a switch on the execution side that would just slow things down. So I guess it is better the way you did it.

janvorli reviewed (approved):
LGTM, thank you!

kg commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -2201,6 +2222,135 @@ bool InterpCompiler::EmitNamedIntrinsicCall(NamedIntrinsic ni, CORINFO_CLASS_HAN
+            int32_t addrVar = m_pStackPointer[0].var;
+            AddIns(INTOP_METHODTABLE);
+            m_pLastNewIns->SetSVar(addrVar);
+            PushStackType(StackTypeO, NULL);
```
Is StackTypeO right? Isn't a methodtable a native pointer not a managed one? I think it should be StackTypeI

janvorli replied:
You are right, I've missed this when reviewing the change, thank you for spotting it!

kotlarmilos replied:
Thanks, changed.

jkotas commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -2360,6 +2360,44 @@ do {                                                                           \
+
+                case INTOP_COMPARE_EXCHANGE_O:
+                {
+                    COMPARE_EXCHANGE(OBJECTREF);
```
CompareExchange for Object argument needs to issue GC write barrier. It should not be expanded as intrinsics, it should just call the FCall that does the right thing. Regular JIT does not expand it as intrinsic either (except for special cases that take null - as an optimization).

kotlarmilos replied:
OK, it now returns false for objects and proceeds with the call. I added a memory barrier after the call. Does it need to be conditional — only if the exchange was successful?

jkotas replied:
You do not need to emit any memory barriers. `InterlockedCompareExchangeT` has the necessary memory barriers.

jkotas replied:
And the FCall with object overload has [all the right GC and memory barriers](https://github.com/dotnet/runtime/blob/135a4af8fde3742ce8fac3c403afffeed791de11/src/coreclr/vm/comutilnative.cpp#L1536) as well.

kotlarmilos committed:
Enhance intrinsic call handling and add null check operation

- Updated EmitNamedIntrinsicCall to include emitMemBarrier parameter.
- Replaced INTOP_METHODTABLE with INTOP_GETMETHODTABLE in EmitNamedIntrinsicCall
- Introduced INTOP_NULLCHECK operation in intops.def and implemented its handling in the interpreter
- Removed debug print statements from compileMethod
- Added tests

kotlarmilos committed:
Replace INTOP_GETMETHODTABLE opcode with INTOP_LDIND_I

jkotas commented on src/coreclr/interpreter/intrinsics.cpp:
```diff
@@ -63,6 +70,15 @@ NamedIntrinsic GetNamedIntrinsic(COMP_HANDLE compHnd, CORINFO_METHOD_HANDLE comp
         // Every intrinsic except IsSupported is PNSE in interpreted-only mode.
         return NI_Throw_PlatformNotSupportedException;
     }
+    else if (HAS_PREFIX(namespaceName, "JIT.HardwareIntrinsics"))
```
JIT.HardwareIntrinsics are tests. Why are recognizing test methods as intrinsics? It does not sound right.

>

jkotas commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -2598,6 +2750,8 @@ void InterpCompiler::EmitCall(CORINFO_RESOLVED_TOKEN* pConstrainedToken, bool re
     int callIFunctionPointerVar = -1;
     void* calliCookie = NULL;
 
+    bool emitMemBarrier = false;
+
```
```suggestion
```
Everything that has to do with emitMemBarrier should be deleted.

kotlarmilos committed:
Remove emitMemBarrier parameter and related logic

kotlarmilos committed:
Use enclosingClassNames to retrieve the top level namespace

kotlarmilos commented on src/coreclr/interpreter/intrinsics.cpp:
```diff
@@ -17,6 +17,13 @@ NamedIntrinsic GetNamedIntrinsic(COMP_HANDLE compHnd, CORINFO_METHOD_HANDLE comp
         return NI_Illegal;
 
+    // Namespace is zero-terinated empty string
+    if (!strcmp(className, "Arm64"))
```
Thanks, I didn't realize it needs to use enclosingClassNames to retrieve the top-level namespace. For this intrinsics a depth of 1 is sufficient, but I’m using 2 (as you suggested) in case deeper nesting is needed.
